d2l-dom.html

```diff
@@ -0,0 +1,62 @@
+<script>
+(function() {
+	'use strict';
+
+	var Dom = {
+
+		getComposedChildren: function(node) {
+
+			if (!node) {
+				return null;
+			}
+			if (node.nodeType !== 1 && node.nodeType !== 9 && node.nodeType !== 11) {
+				return null;
+			}
+
+			var nodes, children = [];
+
+			if (node.tagName === 'CONTENT') {
+				nodes = node.getDistributedNodes();
+			} else {
+				if (node.shadowRoot) {
+					node = node.shadowRoot;
+				}
+				nodes = node.children;
+			}
+
+			for (var i = 0; i < nodes.length; i++) {
+				if (nodes[i].nodeType === 1) {
+					children.push(nodes[i]);
+				}
+			}
+
+			return children;
+
+		},
+
+		getComposedParent: function(node) {
+
+			if (node.getDestinationInsertionPoints) {
+				var insertionPoints = node.getDestinationInsertionPoints();
+				if (insertionPoints && insertionPoints.length > 0) {
+					return insertionPoints[0];
+				}
+			}
+
+			if (node.parentNode) {
+				return node.parentNode;
+			} else if (node.host) {
+				return node.host;
```

Comment:
This is where we walk up the tree from shadow into light-DOM to the host element.